File: problems/p0020_valid_parentheses/solution-0020.java
package p0020_valid_parentheses;

import java.util.Objects;
import java.util.Stack;

import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.Test;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Character[] opening = new Character[] {'(', '{', '['};
        Character[] closing = new Character[] {')', '}', ']'};
        for (char c : s.toCharArray()) {
            for (int i = 0; i < 3; i++) {
                if (c == opening[i]) {
                    stack.push(c);
                } else if (c == closing[i]) {
                    if (stack.isEmpty() || !Objects.equals(stack.pop(), opening[i])) {
                        return false;
                    }
                }
            }
        }
        return stack.isEmpty();

    }
}

class TestCase {
    String s;
    boolean expected;
}

@SuppressWarnings("unused")
class SolutionTest {
    @Test
    void testIsValid() {
        Solution solution = new Solution();
        TestCase[] testCases = new TestCase[] {
            new TestCase() {{
                s = "()";
                expected = true;
            }},
            new TestCase() {{
                s = "()[]{}";
                expected = true;
            }},
            new TestCase() {{
                s = "(]";
                expected = false;
            }},
            new TestCase() {{
                s = "([)]";
                expected = false;
            }},
            new TestCase() {{
                s = "{[]}";
                expected = true;
            }},
            new TestCase() {{
                s = "";
                expected = true;
            }},
            new TestCase() {{
                s = "}";
                expected = false;
            }},
            new TestCase() {{
                s = "((()))";
                expected = true;
            }},
            new TestCase() {{
                s = "[";
                expected = false;
            }},
        };
        for (TestCase testCase : testCases) {
            // Prints test case info for debugging
            System.err.println("s: " + testCase.s);
            System.err.println("expected: " + testCase.expected);
            assertEquals(testCase.expected, solution.isValid(testCase.s));
        }
    }
}

File: problems/p0020_valid_parentheses/solution-0020.py
from dataclasses import dataclass


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {")": "(", "}": "{", "]": "["}
        for char in s:
            if char in mapping:
                top_element = stack.pop() if stack else "#"
                if mapping[char] != top_element:
                    return False
            else:
                stack.append(char)
        return not stack


@dataclass
class TestCase:
    s: str
    expected: bool


def test_sol() -> None:
    sol = Solution()
    cases = [
        TestCase(s="()", expected=True),
        TestCase(s="()[]{}", expected=True),
        TestCase(s="(]", expected=False),
        TestCase(s="([)]", expected=False),
        TestCase(s="{[]}", expected=True),
        TestCase(s="", expected=True),
        TestCase(s="}", expected=False),
        TestCase(s="((()))", expected=True),
        TestCase(s="[", expected=False),
    ]
    for case in cases:
        err_msg = f"Failed for s={case.s}: expected {case.expected}"
        assert sol.isValid(case.s) == case.expected, err_msg
